libchisel/src/funcdeps.rs

```diff
@@ -0,0 +1,83 @@
+use std::collections::HashSet;
+
+use parity_wasm::elements::{CodeSection, Func, FuncBody, FunctionSection, Instruction, Module};
+
+/// A function dependency graph is represented as a list of "edges", or pairs of function indices
+/// (a, b) where a calls b.
+
+/// An edge, where the function at the left index calls the function at the right
+/// index.
+#[derive(PartialEq, Eq, Hash)]
+struct Edge(u32, u32);
+
+/// Container struct for the function dependency graph.
+pub struct DepGraph {
+    edges: HashSet<Edge>,
+}
+
+/// Private interface for managing the function dependency graph
+trait GraphManager {
+    fn probe(&mut self, idx: u32, bodies: &[FuncBody]);
+    fn add_edge(&mut self, dep: Edge) -> bool;
+}
+
+/// Public interface for building function dependency graphs.
+pub trait GraphBuilder: GraphManager {
+    fn build(module: &Module, entry: u32) -> Result<Self, ()>
```

Comment:
Should document `entry` here is a function index.

Reply:
Should also create some helpers outside which translate a function name from `NamesSection` into an index.

Reply:
> Should also create some helpers outside which translate a function name from `Exports`/`NamesSection` into an index.

This is the plan.



> Should document `entry` here is a function index.

fixed. For some reason the trait impl is allowed to have different arg names than the trait definition.